# -*- coding: utf-8 -*-
# OPZIONE 1

import random
from math import fabs
from operator import attrgetter
from deap import creator, base, tools, algorithms
from genetic_algorithm.genetic_evolution import GeneticEvolution
from cut_sequences.selected_dimensional_sequence_numeric import SelectedDimensionalSequenceNumeric
from cut_sequences.dimensional_sequence_binary import DimensionalSequenceBinary
from matplotlib import pyplot as plt
import numpy as np


# Class for utilization of genetic guide using DEAP
class DeapGeneticGuideSequenceProblem(GeneticEvolution):

    # Definition constructor class
    # @individual_size: number of attributes for each individual
    # @mutation_rate: percentage of mutation
    # @mating_rate: mating ratio
    # @selected_individuals: number of individuals selected in the phase of selection of the best individuals
    # (selection by tournament)
    # @cuts_sequence: sequence of cuts T_d
    # @points_list: list of prototypes
    # @elements_per_dimension: number of elements per dimension of given cuts_sequence for convertion from chromosome
    # to sequence
    # @min_cut: m_d cut that will go into S_d
    # @max_cut: M_d cut that will go into S_d
    def __init__(self, individual_size, mutation_rate, mating_rate, selected_for_tournament, cuts_sequence, points_list,
                 elements_per_dimension, min_cut, max_cut):

        # initialize the seed for random numbers
        random.seed()

        # define mutation rate of individual
        self.mutation_rate = mutation_rate

        # define mating rate of individual
        self.mating_rate = mating_rate

        # define how many individuals are to be selected for tournament
        self.selected_for_tournament = selected_for_tournament

        # create object defining max fitness value
        creator.create("FitnessMax", base.Fitness, weights=(1.0,))

        # create object defining individual
        creator.create("Individual", list, fitness=creator.FitnessMax)

        # define toolbox for needed methods
        self.toolbox = base.Toolbox()

        # define individual, which will be generated by a given "generate_fun" function
        self.toolbox.register("individual", self.generate, creator.Individual, individual_size)

        # define data structure "population" containing all individuals
        self.toolbox.register("population", tools.initRepeat, list, self.toolbox.individual)

        # define mating method between individuals using uniform partially matched crossover
        self.toolbox.register("mate", tools.cxUniformPartialyMatched, indpb=mating_rate)

        # define mutation method of individuals' son shuffling genes with "mutation_rate" percentage
        self.toolbox.register("mutate", tools.mutFlipBit, indpb=mutation_rate)

        self.toolbox.register("evaluate", self.evaluate)

        # define selection method using selection for tournament between "selected_for_tournament" individuals
        # TODO - "relazione a distanza", test
        self.toolbox.register("select", tools.selTournament, tournsize=selected_for_tournament)
        # TODO - "vicini di casa", test
        # self.toolbox.register("select", self.tournament_selection_1_tournament, tournsize=selected_for_tournament)

        # save cuts_sequence, points_list, individual_size and element_per_dimension, needed for convertion from
        # chromosome to sequence
        self.T_d = cuts_sequence
        self.points_list = points_list
        self.individual_size = individual_size
        self.elements_per_dimension = elements_per_dimension
        # save m_d and M_d limits to generate S_d sequence
        self.m_d = min_cut
        self.M_d = max_cut

    # Function that evaluates the ratio between the "true" genes (considered cuts) and the whole number of them
    # (considered cuts and not)
    # @individual: object that contains the genome
    def evaluate(self, individual):
        # initializing evaluation variables
        valutation = 0
        total_genes = 0
        # every "true" gene increments the valutation while the total number of genes is calculated
        valutation += individual.count(True)
        total_genes += len(individual)
        # returns the ratio
        return valutation / total_genes

    # Function that generates an individual with the same number of cuts as the cuts sequence
    # @individual_class: class of the individual to create
    # @individual_dim: number of genes of the individual
    def generate(self, individual_class, individual_dim):
        # definition of individual's genome
        genome = list()

        # initializing the genome with all genes to False
        for gene in range(individual_dim):
            genome.append(False)

        # return the individual with the created genome
        return individual_class(genome)

    # Method generating the best individual possible by the genetic algorithm
    # @population_size: size of population to generate
    # @generations: number of generations to create
    # @selected_best: number of best individuals to generate
    def evolve(self, population_size, generations, dataset):

        # create a population of "population_size" individuals
        population = self.toolbox.population(n=population_size)

        fit_behave = list(tuple())  # TODO - valutazione fitness, da togliere
        bestfit = 0
        bestind = None

        epoch = 0
        stabilized_gens = 0
        # previous_avg_fit = 0

        # for each generation
        # for epoch in range(generations - 1):
        # TODO - aggiunto altro criterio di stop "stabilizzazione", test
        while epoch in range(generations - 1) and stabilized_gens < 10:

            # offsprings are generated using the offsprings_generator method, in which are passed the population,
            # population_size, toolbox, mating rate and mutation rate
            offsprings = self.offsprings_generator(population, population_size, self.toolbox, self.mating_rate,
                                                   self.mutation_rate)
            # offsprings = algorithms.varAnd(population, self.toolbox, self.mating_rate, self.mutation_rate)

            # definizione di una mappa che conterrÃ  i valori delle
            # valutazioni degli individui della progenie
            fitness = self.toolbox.map(self.fitness, offsprings)

            # TODO - valutazione fitness, da togliere
            eval_fitness = list()
            for son in offsprings:
                eval_fitness.append(self.fitness(son))
            current_max_fit = 0
            current_min_fit = 1
            temp_avg = 0
            for i in range(len(eval_fitness)):
                if eval_fitness[i] > current_max_fit:
                    current_max_fit = eval_fitness[i]
                if eval_fitness[i] < current_min_fit:
                    current_min_fit = eval_fitness[i]
                temp_avg += eval_fitness[i]
                current_avg_fit = temp_avg / population_size
            fit_behave.append((current_min_fit, current_avg_fit, current_max_fit))

            # TODO - aggiunto altro criterio di stop "stabilizzazione", test
            # if fabs(previous_avg_fit - current_avg_fit) <= previous_avg_fit * 0.1:
            #     stabilized_gens += 1
            # else:
            #     stabilized_gens = 0
            # previous_avg_fit = current_avg_fit

            # map each fitness value to the corresponding offspring
            for fit, ind in zip(fitness, offsprings):
                # if an individual had better fitness than the best found so far
                if fit > bestfit:
                    # save the better individual
                    bestfit = fit
                    bestind = ind
                ind.fitness.value = fit

            # select a couple of offsprings in the population that will be "mother" and "father" of the next batch of
            # generated individuals
            # the selection is defined on a "selected_for_tournament" number of offsprings
            # population = self.toolbox.select(offsprings, k=2)  # TODO - "famiglia tradizionale", test
            # population = self.toolbox.select(offsprings, k=self.selected_for_tournament)  # TODO - "bisbocce", test
            population = self.toolbox.select(offsprings, k=int(population_size))  # TODO - "assembramento", da test

            epoch += 1

        # TODO - grafico valutazione fitness, da togliere
        min_ = list()
        avg_ = list()
        max_ = list()
        for fits in fit_behave:
            min_.append(fits[0])
            avg_.append(fits[1])
            max_.append(fits[2])
        x = np.linspace(0, len(fit_behave), len(fit_behave))
        plt.plot(x, min_, marker='.', color='red')
        plt.plot(x, avg_, marker='.', color='green')
        plt.plot(x, max_, marker='.', color='blue')
        plt.grid(True)
        # plt.show()
        plt.savefig(dataset + ".svg", transparent=True)
        plt.close()

        print("Best fitness: ", bestfit)  # TODO - valutazione fitness, da togliere
        print("Halt at generation:", epoch)  # TODO - doppio criterio di fermata "evolve", da togliere
        # convert individual into sequence
        best_individual = self.from_list_to_sequence(bestind, self.elements_per_dimension)

        # return the converted best individual
        # return best_individual
        return bestind.count(True), self.individual_size

    # Method defining the fitness value of an individual
    # @individual: individual
    def fitness(self, individual):
        # return the calculated fitness value
        # return (1 - self.toolbox.evaluate(individual)) * pow(self.pureness(individual), 5)
        return self.toolbox.evaluate(individual)

    # Method that calculates the pureness of a given individual's genome
    # @individual: individual
    def pureness(self, individual):
        # initializing selected cuts and binary cuts sequences
        S_d = SelectedDimensionalSequenceNumeric()
        S_d_b = DimensionalSequenceBinary()

        # convert individual into sequence
        converted_individual = self.from_list_to_sequence(individual, self.elements_per_dimension)

        # create binary cuts sequence
        S_d_b.from_binary(converted_individual)
        # generate selected cuts sequence from reference T_d and binary cuts sequence
        S_d.from_binary(self.T_d, S_d_b)

        # create set of hyperboxes
        hyperboxes = S_d.generate_hyperboxes_set(self.points_list, self.m_d, self.M_d)

        # return ratio between number of pure hyperboxes and total number of hyperboxes
        return hyperboxes.get_pure_hyperboxes_number() / hyperboxes.get_hyperboxes_number()

    # Method converting individual's genome from list to "multidimensional cuts sequence" (list of lists)
    # @individual: individual
    # @genes_per_dimension: numbers of genes per dimension
    def from_list_to_sequence(self, individual, genes_per_dimension):
        # create support lists
        sequence = list()
        dimension = list()

        # initialize index and offset variables
        offset = 0
        i = 0
        # for each dimension
        for num_elem in genes_per_dimension:
            # clear the "dimension" support list
            dimension.clear()
            # increment the offset that limits how many genes are copied into given dimension
            offset = offset + num_elem

            # while index doesn't reach the max elements in given dimension
            while i < offset:
                # append into given dimension the gene with given index from individual
                dimension.append(individual[i])
                # increment index
                i += 1
            # append a copy of the newly created dimension into sequence
            sequence.append(dimension.copy())
        return sequence

    # Method that generates a offspring population of population_size using "mate" and "mutate" methods
    # @population: list of individuals
    # @population_size: size of population to generate
    # @toolbox: container of utility functions
    # @mating_rate: mating ratio
    # @mutation_rate: mutation_ratio
    def offsprings_generator(self, population, population_size, toolbox, mating_rate, mutation_rate):
        clones = [toolbox.clone(ind) for ind in population]
        offsprings = list()
        dummy_ind = max(clones, key=attrgetter("fitness"))
        clones.sort(key=lambda fitness: fitness, reverse=False)
        for _ in range(population_size):
            offsprings.append(dummy_ind)
        # Apply crossover and mutation on the offsprings
        for i in range(2, population_size):
            if random.random() < mating_rate:
                offsprings[i - 1], offsprings[i] = \
                        toolbox.mate(clones[random.randint(0, len(population) - 1)],
                                     clones[random.randint(0, len(population) - 1)])

        for i in range(1, population_size):
            if random.random() < mutation_rate:
                offsprings[i], = toolbox.mutate(offsprings[i])

        return offsprings
